core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/utils/InfraUtils.kt

```diff
@@ -20,3 +17,149 @@ fun TrackNetworkInfra.getNextTrackSections(
+        }
+    }
+    return res
+}
+
+/** Returns the blocks that follow the given one */
+fun BlockInfra.getNextBlocks(infra: RawSignalingInfra, blockId: BlockId): Set<BlockId> {
+    val entry = getBlockExit(infra, blockId)
+    return getBlocksStartingAtDetector(entry).toSet()
+}
+
+/** Returns the block's entry detector */
+fun BlockInfra.getBlockEntry(rawInfra: RawInfra, block: BlockId): DirDetectorId {
+    val blockPath: StaticIdxList<ZonePath> = getBlockPath(block)
+    val firstZone: ZonePathId = blockPath[0]
+    return rawInfra.getZonePathEntry(firstZone)
+}
+
+/** Returns the block's exit detector */
+fun BlockInfra.getBlockExit(rawInfra: RawInfra, block: BlockId): DirDetectorId {
+    val blockPath: StaticIdxList<ZonePath> = getBlockPath(block)
+    val lastZonePath: ZonePathId = blockPath[blockPath.size - 1]
+    return rawInfra.getZonePathExit(lastZonePath)
+}
+
+/** Returns the blocks that lead into the given one */
+fun BlockInfra.getPreviousBlocks(infra: RawSignalingInfra, blockId: BlockId): Set<BlockId> {
+    val entry = getBlockEntry(infra, blockId)
+    return getBlocksEndingAtDetector(entry).toSet()
+}
+
+/** Returns the route's corresponding blocks */
+fun BlockInfra.getRouteBlocks(
+    rawInfra: RawInfra,
+    route: RouteId,
+    allowedSigSystems: StaticIdxList<SignalingSystem>? = null
+): StaticIdxList<Block> {
+    val blockPaths =
+        recoverBlocks(rawInfra, this, mutableStaticIdxArrayListOf(route), allowedSigSystems)
+    if (blockPaths.isEmpty()) return mutableStaticIdxArrayListOf()
+    // No signaling system for now, take the first block path possibility.
+    // Correct when signalisation is taken into account.
+    val blocks = blockPaths[0].toBlockList()
+    return blocks
+}
+
+/** Finds a valid route that follows the given path */
+private fun findRoute(
+    infra: RawSignalingInfra,
+    blockInfra: BlockInfra,
+    chunks: DirStaticIdxList<TrackChunk>,
+    startIndex: Int,
+    routeMustIncludeStart: Boolean
+): RouteId {
+    val routes = infra.getRoutesOnTrackChunk(chunks[startIndex])
+
+    // We need to evaluate the longest route first, in case one route covers a subset of another
+    val sortedRoutes = routes.sortedBy { r -> -infra.getChunksOnRoute(r).size }
+    for (routeId in sortedRoutes) if (
+        routeMatchPath(infra, blockInfra, chunks, startIndex, routeMustIncludeStart, routeId)
+    )
+        return routeId
+    throw RuntimeException("Couldn't find a route matching the given chunk list")
+}
+
+/** Returns false if the route differs from the path */
+private fun routeMatchPath(
+    infra: RawSignalingInfra,
+    blockInfra: BlockInfra,
+    chunks: DirStaticIdxList<TrackChunk>,
+    chunkIndex: Int,
+    routeMustIncludeStart: Boolean,
+    routeId: RouteId
+): Boolean {
+    var mutChunkIndex = chunkIndex
+    if (!routeHasBlockPath(infra, blockInfra, routeId))
+        return false // Filter out routes that don't have block, they would cause issues later on
```

Comment:
I know that's not quite related, but I don't know if I'm comfortable having this check here: what does "later on" mean? sure, in some contexts, a route may need to have both blocks and match a given path. But then this is best checked in this specific context.

Reply:
"later on" would mean "any time we call `recoverBlocks` and assert that there is a result", which is quite often (and not actually STDCM-related, this has just been moved). Most of the code has been written with the assumption that every route has a block path. 

In my opinion these routes should be discarded as part of the loading process and not part of the infra, but I don't know how that would be implemented (considering the blocks are loaded after the routes)

Reply:
> Most of the code has been written with the assumption that every route has a block path

I think that's the core issue here. Because not all blocks types can be used by all trains, I don't see how that's relevant: even if a route is entirely covered by blocks, they may not be usable

Reply:
I'd say fixing this is out of scope here, it's not directly related and I think it would be a fairly large change. I can open an issue.

Reply:
https://github.com/osrd-project/osrd/issues/6884